pybamm/expression_tree/binary_operators.py

```diff
@@ -573,17 +592,25 @@ def _binary_evaluate(self, left, right):
     def _binary_simplify(self, left, right):
         """ See :meth:`pybamm.BinaryOperator.simplify()`. """
 
-        # anything multiplied by a scalar zero returns a scalar zero
+        # simplify multiply by scalar zero, being careful about shape
         if is_scalar_zero(left):
-            if isinstance(right, pybamm.Array):
-                return pybamm.Array(np.zeros(right.shape))
-            else:
+            shape = right.shape_for_testing
+            if shape == ():
                 return pybamm.Scalar(0)
-        if is_scalar_zero(right):
-            if isinstance(left, pybamm.Array):
-                return pybamm.Array(np.zeros(left.shape))
             else:
+                if len(shape) == 1 or shape[1] == 1:
+                    return pybamm.Vector(np.zeros(shape))
+                else:
+                    return pybamm.Matrix(csr_matrix(shape))
+        if is_scalar_zero(right):
+            shape = left.shape_for_testing
+            if shape == ():
                 return pybamm.Scalar(0)
+            else:
+                if len(shape) == 1 or shape[1] == 1:
+                    return pybamm.Vector(np.zeros(shape))
+                else:
+                    return pybamm.Matrix(csr_matrix(shape))
```

Comment:
seeing them all together in review mode, these lines appear several times together and could be abstracted into a functon `pybamm.zeros_of_shape(shape)`

Reply:
good suggestion